File: problems/1031-maximum-sum-of-two-non-overlapping-subarrays/main.py
from typing import List
from itertools import accumulate

# ref: https://leetcode.cn/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/2245647/tu-jie-mei-you-si-lu-yi-zhang-tu-miao-do-3lli/


class Solution:
    def maxSumTwoNoOverlap(self, nums: List[int], firstLen: int, secondLen: int) -> int:
        s = list(accumulate(nums, initial=0))
        ans = 0

        def help(firstLen: int, secondLen: int) -> None:
            nonlocal ans
            maxSumA = 0
            # traverse B index
            for i in range(firstLen+secondLen, len(s)):
                # compare sum of previous A with the array ahead of B
                maxSumA = max(maxSumA, s[i-secondLen]-s[i-secondLen-firstLen])
                # compare ans with sum of A and B
                ans = max(ans, maxSumA+s[i]-s[i-secondLen])
        # change the order of firstLen and secondLen to go through all cases
        help(firstLen, secondLen)
        help(secondLen, firstLen)

        return ans

File: problems/1031-maximum-sum-of-two-non-overlapping-subarrays/main.cpp
#include <iostream>
#include <numeric>
#include <ostream>
#include <vector>

class Solution {
public:
  int maxSumTwoNoOverlap(std::vector<int> &nums, int firstLen, int secondLen) {
    std::vector<int> s(nums.size() + 1, 0);
    std::partial_sum(nums.begin(), nums.end(),
                     s.begin() + 1); // Calculate prefix sums
    int ans = 0;

    auto help = [&](int L1, int L2) -> void {
      int maxSumA = 0;
      for (int i = L1 + L2; i < s.size(); ++i) {
        maxSumA = std::max(maxSumA, s[i - L2] - s[i - L2 - L1]);
        ans = std::max(ans, maxSumA + s[i] - s[i - L2]);
      }
    };

    help(firstLen, secondLen);
    help(secondLen, firstLen);
    return ans;
  }
};

int main() {
  std::vector<int> nums = {0, 6, 5, 2, 2, 5, 1, 9, 4};
  int firstLen = 1;
  int secondLen = 2;
  std::cout << Solution().maxSumTwoNoOverlap(nums, firstLen, secondLen)
            << std::endl;

  return 0;
}